File: List and Node Manipulation (Java)/Lex.java
// Student: Josh, Tsung Chi, Shih
// CruzID/CruzID#: jtshih/1436140
// Class: CMPS 101
// Date: 1/14/2015
// Filename: Lex.java
// Description: Takes two arguments, inputFile and outputFile, respectively. Reads inputFile line by line and
//				makes/overwrites an outputFile of the inputFile that is Alphabetically sorted.

import java.io.*;
import java.util.Scanner;


public class Lex {
    static void processFile(String inFile, String outFile) throws IOException { //plain read file and put contents into ADT
    		
    	int numOfLines = 0;
    	//read file, put contents Alphabetically into ADT, then write to another file
    	Scanner scan = new Scanner (new File(inFile));
    	Scanner lineNum = new Scanner (new File(inFile));
    	PrintWriter out = new PrintWriter(new FileWriter(outFile));
        List list = new List();
        while(lineNum.hasNextLine())
        {
        	numOfLines++;
        	lineNum.nextLine();
        }
	    lineNum.close();
	    
    	String[] strArray = new String[numOfLines];
        for (int i = 0; i<numOfLines; i++) 
            strArray[i] = scan.nextLine();
        
	    list.append(0);
	    for(int i=1; i<numOfLines; i++)
	    {
	    	list.moveTo(0);
	    	for(;;)
	    	{
	    		if(strArray[i].compareTo(strArray[list.getIndex()]) < 0) //goes through entire list
		    	{
		    		if(list.getIndex()==0)
		    			list.prepend(i);
		    		else
		    			list.insertBefore(i);
		    		break;
		    	}
	    		else
	    		{
	    			list.moveNext();
	    			if(list.getIndex() == -1)
	    			{
	    				list.append(i);
	    				break;
	    			}
	    		}
	    	}
	    }

	    for(int i=0; i<numOfLines; i++)
	    {
	    	list.moveTo(i);
	    	out.println(strArray[list.getElement()]);
	    	System.out.print(strArray[list.getElement()]+" ");
	    }
	    System.out.println();
	    
		scan.close();
		out.close();
    }
    public static void main(String[] args) {
//    	String[] targs = new String[2];
//    	args[0] = "in.txt";
//    	args[1] = "out.txt";
	
        String inputFile = "";
	try{
		inputFile = args[0];
	}catch (ArrayIndexOutOfBoundsException e){
         throw new RuntimeException("No input file detected");
      }
        String outputFile = "";
	try{
		outputFile = args[1];
	}catch (ArrayIndexOutOfBoundsException e){
         throw new RuntimeException("No output file detected");}

        try{
        	if(args[2] != null)
        	throw new IllegalArgumentException("Error! Too many arguments");
        }catch (ArrayIndexOutOfBoundsException e){}
        
        
        try {
			processFile(inputFile, outputFile);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }
}

File: List and Node Manipulation (C)/Lex.c
// Student: Josh, Tsung Chi, Shih
// CruzID/CruzID#: jtshih/1436140
// Class: CMPS 101
// Date: 1/23/2015
// Filename: Lex.c

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 160

int main(int argc, char * argv[]){

   int count=0;
   FILE *in, *out;
   char line[MAX_LEN];
   char* token;
   char hold[MAX_LEN][MAX_LEN];
   List list = newList();

   // check command line for correct number of arguments
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   /* read each line of input file, then count and print tokens */
   while( fgets(line, MAX_LEN, in) != NULL)  {
      count++;
      token = strtok(line, " \n");
      strcpy(hold[count-1], token);
   }
   hold[count][0] = '\0';
   append(list, 0);
      for(int i=1; i<count; i++)
	{
	    	moveTo(list, 0);
	    	for(;;)
	    	{
	    		if(strcmp(hold[i], hold[getElement(list)]) < 0) //goes through entire list
		    	{
		    		if(getIndex(list)==0)
		    			prepend(list, i);
		    		else
		    			insertBefore(list, i);
		    		break;
		    	}
	    		else
	    		{
	    			moveNext(list);
	    			if(getIndex(list) == -1)
	    			{
	    				append(list, i);
	    				break;
	    			}
	    		}
	    	}
	    }

   for(int i=0; i<count; i++){
      moveTo(list, i);
      fprintf(out, "%s\n", hold[getElement(list)]);
   }
   printList(stdout, list);
   printf("\n");
   /* close files */
   fclose(in);
   fclose(out);

   return(0);
}